### PR TITLE
refactor(sequencer): generate `SequencerBlock` after transaction execution in proposal phase

### PR DESCRIPTION
**Pull request by noot**
## Summary
refactor the sequencer app to generate and store the resulting `SequencerBlock` after transaction execution even in the proposal phase. 

## Background
previously, we were only generating the `SequencerBlock` in `finalize_block`, however with the upcoming builder APIs (#1519) we require the (proposed) `SequencerBlock` to be available after execution in the proposal phase. 

## Changes
- create a `post_execute_transactions` method and move the after-execution logic that generates the `SequencerBlock` from `finalize_block` to there.
- call this method after transaction execution in `process_proposal`. 
- if txs were executed in `prepare_proposal`, `post_execute_transactions` is still called in `process_proposal`, as the block hash is not available in `prepare_proposal`.

## Testing
existing unit tests pass, app logic was not changed, just moved around

## Related Issues
 related to #1322

### CODE REVIEW

File: crates/astria-sequencer/src/app/mod.rs [hunk 1]
```diff
@@ -162,9 +162,13 @@ pub(crate) struct App {
...
-    // cache of results of executing of transactions in `prepare_proposal` or `process_proposal`.
-    // cleared at the end of each block.
-    execution_results: Option<Vec<tendermint::abci::types::ExecTxResult>>,
+    // cache of results of executing of transactions in `prepare_proposal`.
+    // cleared in `process_proposal` if we're the proposer.
+    execution_results: Option<Vec<ExecTxResult>>,
+
+    // cache of results of executing of transactions in `process_proposal`.
+    // cleared at the end of the block.
+    finalize_block: Option<abci::response::FinalizeBlock>,
```
Comment (joroshiba) [inline crates/astria-sequencer/src/app/mod.rs:171]: Should we utilize the ephemeral storage within cnidarium vs adding to state on the app? I know @SuperFluffy has been trying to decrease our global state bloat.

Comment (noot) [reply crates/astria-sequencer/src/app/mod.rs:171]: this is functionally the same - it's an object stored in memory and cleared at the start of each block. the ephemeral store is also only in memory and is cleared at the end of each block

Comment (SuperFluffy) [reply crates/astria-sequencer/src/app/mod.rs:171]: I agree that we should employ the ephemeral store more rather than adding fields to app.

There is a tendency to add more and more fields to `App` that need to be actively set/unset, which is not clean (and I believe with which we had issues in the past).

For a reader it's also better to have a single source of truth of effects that happen during block building, rather than finding some in the ephemeral store and some in the "global" app state.

File: crates/astria-sequencer/src/app/mod.rs [hunk 2]
```diff
@@ -458,6 +482,16 @@ impl App {
         );
 
         self.executed_proposal_hash = process_proposal.hash;
+        self.post_execute_transactions(
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/app/mod.rs:485]: Why not pass this the entire `process_proposal`?

File: crates/astria-sequencer/src/app/tests_app/mempool.rs
Comment (SuperFluffy) [inline crates/astria-sequencer/src/app/tests_app/mempool.rs:1]: This test is so complex, I don't know how to evaluate this change to it.

File: crates/astria-sequencer/src/app/mod.rs [hunk 3]
```diff
@@ -758,18 +791,24 @@ impl App {
...
+    /// `SequencerBlock`.
     ///
-    /// This is called by cometbft after the block has already been
-    /// committed by the network's consensus.
-    #[instrument(name = "App::finalize_block", skip_all)]
-    pub(crate) async fn finalize_block(
+    /// this must be called after a block's transactions are executed.
+    #[instrument(name = "App::post_execute_transactions", skip_all)]
+    async fn post_execute_transactions(
         &mut self,
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/app/mod.rs:800]: Hard to follow all the side effects in this function.

File: crates/astria-sequencer/src/app/mod.rs [hunk 4]
```diff
@@ -781,16 +820,73 @@ impl App {
...
+            .wrap_err("failed to write sequencer block to state")?;
+
+        // events that occur after end_block are ignored here;
+        // there should be none anyways.
+        let _ = self.apply(state_tx);
+
+        // use a dummy app hash here - the actual app hash will be filled out in finalize_block.
+        let dummy_app_hash = AppHash::default();
+
+        let finalize_block = abci::response::FinalizeBlock {
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/app/mod.rs:867]: It's surprising that this method changes the internal state of `self.finalize_block`, but the `finalize_block` takes a `finalize_block` argument and then mutates the `finalize_block` field that some other part of the stack changed.

**Review by SuperFluffy — COMMENTED** [review]
I have a hard time following the changes, which might be an artifact of how app execution is structured.

I am also surprised that making sequencer blocks available during an earlier phase leads to changes in mempool tests.

File: crates/astria-sequencer/src/app/mod.rs (hunk 2 above)
Comment (noot) [reply crates/astria-sequencer/src/app/mod.rs:485]: this is also called in `finalize_block` potentially, so it might not have `process_proposal` available

File: crates/astria-sequencer/src/app/tests_app/mempool.rs
Comment (noot) [reply crates/astria-sequencer/src/app/tests_app/mempool.rs:1]: yeah :/ i didn't write the test, just modified it/added some asserts for sanity when updating it

File: crates/astria-sequencer/src/app/mod.rs (hunk 4 above)
Comment (noot) [reply crates/astria-sequencer/src/app/mod.rs:867]: i could change this to be some intermediate type that doesn't have the app hash yet, which is turned into `finalize_block` response only in `finalize_block`. would that be easier to follow?

**Comment by noot** [comment]
> I am also surprised that making sequencer blocks available during an earlier phase leads to changes in mempool tests.

the changes are because `process_proposal` must now also be called if `prepare_proposal` is called (this is the normal ABCI flow btw). that's really the only change to the tests

File: crates/astria-sequencer/src/app/mod.rs (hunk 3 above)
Comment (noot) [reply crates/astria-sequencer/src/app/mod.rs:800]: this function is basically everything that was in `finalize_block` after txs were executed (minus mempool cleaning, as that needs to be done only when the block is actually finalized)

File: crates/astria-sequencer/src/app/mod.rs (hunk 1 above)
Comment (noot) [reply crates/astria-sequencer/src/app/mod.rs]: moved the fields to the ephemeral object store here: https://github.com/astriaorg/astria/pull/1562/commits/591fd94c89ae27b26c5b067f423dd2a657f87d41

this is nice because the fields are automatically cleared when `update_state_for_new_round` is called, however having to get a `StateDelta` for writing to it is bit more cumbersome in parts (putting the `execution_results` in `prepare_proposal`)

**Review by Fraser999 — APPROVED** [review]
There's one mention of `self.execution_results` left behind in the doc comments for `execute_transactions_prepare_proposal` which would be good to have updated.

I also think we should look to abstract the usage of the ephemeral store behind our own API provided by the extension traits in `app/state_ext.rs` (similar to how cached block deposits are handled in the `bridge` extension traits).  We'd also then keep the keys for these in `app::state_ext` along with the others.  But I can handle changing to that as part of the ongoing storage updates.

Neither of these are blocking though, so approving as is.

File: crates/astria-sequencer/src/app/mod.rs [hunk 5]
```diff
@@ -629,15 +662,16 @@ impl App {
         self.metrics
             .set_transactions_in_mempool_total(self.mempool.len().await);
 
-        self.execution_results = Some(execution_results);
+        let mut state_tx = Arc::try_begin_transaction(&mut self.state)
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/app/mod.rs:665]: This looks complex. Can you add comment above explaining why it's necessary? Seems like a hack around cnidarium not providing us with something we need.

File: crates/astria-sequencer/src/app/mod.rs (hunk 4 above)
Comment (SuperFluffy) [reply crates/astria-sequencer/src/app/mod.rs]: I suppose it's fine. We should probably sit down and refactor of all of this. It looks like we have a bit of code duplication going on.
